static/js/components/Periodogram.jsx

```diff
@@ -0,0 +1,549 @@
+  paper: {
+    width: "100%",
+    padding: theme.spacing(1),
+    textAlign: "left",
+    color: theme.palette.text.primary,
+  },
+  items: {
+    maxWidth: "100%",
+    width: "10rem",
+    fullWidth: "true",
+    display: "flex",
+    wrap: "nowrap",
+  },
+  spinner: {
+    position: "relative",
+    margin: "auto",
+    width: "50%",
+    fontWeight: "bold",
+    fontSize: "1.25rem",
+    textAlign: "center",
+  },
+  labels: {
+    fontSize: "0.75rem",
+  },
+}));
+
+function transpose(mat) {
+  // transpose matrix
+  return mat[0].map((col, i) => mat.map((row) => row[i]));
+}
+
+function plotdata(xx, yy, me) {
+  const dat = transpose([xx, yy])
+    .map((x, i) => [x[0], [x[1], me[i]]])
+    .sort((a, b) => b[0] - a[0]);
+  // eslint-disable-next-line no-new
+  new Dygraph(document.getElementById("dataplot"), dat, {
+    drawPoints: true,
+    strokeWidth: 0,
+    labels: ["time", "mag"],
+    errorBars: true,
+    valueRange: [Math.max(...yy) + 0.1, Math.min(...yy) - 0.1],
+    dateWindow: [Math.min(...xx) - 6, Math.max(...xx) + 6],
+  });
+}
+
+function plotphased(xx, yy, p, title) {
+  // Create graph with native array as data source
+  const pp = [...xx.map((x) => x % p), ...xx.map((x) => (x % p) + p)];
+  // eslint-disable-next-line no-new
+  new Dygraph(
+    document.getElementById("phased"),
+    transpose([pp, [...yy, ...yy]]),
+    {
+      drawPoints: true,
+      strokeWidth: 0,
+      labels: ["phase", "mag"],
+      valueRange: [Math.max(...yy) + 0.1, Math.min(...yy) - 0.1],
+      dateWindow: [0, 2 * p],
+      title,
+    }
+  );
+}
+
+function plotline(g, x) {
+  const lines = [
+    [1.0, "rgb(255,0,0,0.8)", 3],
+    [0.5, "rgb(200,50,50,0.2)", 1],
+    [3.0, "rgb(200,50,50,0.2)", 1],
+    [2.0, "rgb(200,50,50,0.2)", 1],
+    [4.0, "rgb(200,50,50,0.2)", 1],
+  ];
+
+  g.updateOptions({
+    underlayCallback(canvas) {
+      for (let i = 0; i < lines.length; i += 1) {
+        const loc = Math.log10(x * lines[i][0]);
+        canvas.beginPath();
+        canvas.moveTo(g.toDomXCoord(loc), g.toDomYCoord(0));
+        canvas.lineTo(g.toDomXCoord(loc), g.toDomYCoord(10));
+        [canvas.strokeStyle, canvas.lineWidth] = [lines[i][1], lines[i][2]];
+        canvas.stroke();
+      }
+    },
+  });
+}
+
+function dot(x, y) {
+  let i = x.length;
+  let sum = 0;
+  // eslint-disable-next-line no-plusplus
+  while (i--) sum += x[i] * y[i];
+  return sum;
+}
+
+function add(x, a) {
+  let i = x.length;
+  const xa = [];
+  // eslint-disable-next-line no-plusplus
+  while (i--) xa[i] = x[i] + a;
+  return xa;
+}
+
+function mul(x, y) {
+  let i = y.length;
+  const xy = [];
+  if (x.length === i) {
+    // eslint-disable-next-line no-plusplus
+    while (i--) xy[i] = x[i] * y[i];
+  } else {
+    // eslint-disable-next-line no-plusplus
+    while (i--) xy[i] = x * y[i];
+  }
+  return xy;
+}
+
+// Generalised Lomb-Scargle periodogram: https://github.com/mzechmeister/GLS/tree/master/javascript
+function GLS(t_data, y_data, kwa) {
```

Comment:
Is the plan to eventually replace this with an N log N algorithm such as Press & Rybicki, or will we get satisfactory performance with N^2 for most curves?

Reply:
Plan is to use this until we push to a graphics card implementation.